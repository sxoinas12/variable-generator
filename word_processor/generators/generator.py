import spacy
from py.error import Error

nlp = spacy.load("en_core_web_trf")


class Generator():
    """
        Generator class that will use a strategy resolver to generate suggestions for a text
        :return:

    """

    # passed text
    text = None
    # parsed text from spacy lib
    doc = None
    # the strategy resolver to use
    strategy_resolver = None
    # the parser for the presentation
    parser = None

    # suggestions calculated
    suggestions = None

    def __init__(self, text=None, strategy=None, parser=None):
        self.strategy_resolver = strategy
        self.parser = parser
        if text is not None:
            self.set_text(text)

    def suggest(self):
        if self.doc is None or self.text is None:
            raise Error('No document')

        self.suggestions = self.strategy_resolver(self.doc)

        return self.suggestions

    def set_text(self, text):
        self.text = text
        self.doc = nlp(text)

    def set_strategy(self, strategy_resolver):
        self.strategy_resolver = strategy_resolver

    def present(self):
        self.parser.suggestions = self.suggestions
        return self.parser.present()