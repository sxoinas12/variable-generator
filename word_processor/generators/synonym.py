import nltk
from nltk.corpus import wordnet
import os

from nltk.corpus.reader import WordNetError

print("Loading nltk", os.path.dirname(__file__) + '/../../nltk_data/')
nltk.data.path.append(os.path.dirname(__file__) + '/../../nltk_data/')


def get_synonyms(word):
    synonyms = []

    try:
        synsets_words = wordnet.synsets(word)
    except WordNetError:
        return []

    for syn in synsets_words:
        for lemma in syn.lemmas():
            synonyms.append(lemma.name())

    return synonyms


def get_antonyms(word):
    antonyms = []

    try:
        synsets_words = wordnet.synsets(word)
    except WordNetError:
        return []

    for syn in synsets_words:
        for lemma in syn.lemmas():
            if lemma.antonyms():
                for antonym in lemma.antonyms():
                    antonyms.append(antonym.name())

    return antonyms