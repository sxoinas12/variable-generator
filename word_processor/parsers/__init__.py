from .parser import Parser
from .javascript import JavascriptParser

__all__ = [
    'Parser',
    'JavascriptParser',
]