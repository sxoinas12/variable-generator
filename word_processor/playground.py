import spacy
from spacy import displacy

nlp = spacy.load("en_core_web_trf")


def test_something(text):
    doc = nlp(text)

    displacy.serve(doc, style="dep", port=5001)


test_something('allow chat communication for authenticated users')
#test_something('Stick element after scroll has passed the element')