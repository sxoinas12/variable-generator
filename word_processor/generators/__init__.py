from .generator import Generator
from .synonym import get_synonyms, get_antonyms

__all__ = [
    'Generator',
    'get_antonyms',
    'get_synonyms',
]
