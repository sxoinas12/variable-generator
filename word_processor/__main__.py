import sys

from word_processor.generators import Generator
from word_processor.parsers import JavascriptParser
from word_processor.strategies import back_and_forth_strategy


def calculate_names(text: str):
    generator = Generator(text=text, strategy=back_and_forth_strategy, parser=JavascriptParser())

    generator.suggest()

    presentation = generator.present()

    print('\n\n'.join(presentation))


# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    calculate_names(sys.argv[1])

# See PyCharm help at https://www.jetbrains.com/help/pycharm/
