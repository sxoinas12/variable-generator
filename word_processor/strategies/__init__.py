from .leaf_strategy import leaf_strategy
from .leaf_no_preps_strategy import leaf_no_prep_strategy
from .adverbial_strategy import adverbial_strategy
from .compound_strategy import compound_strategy
from .flip_prep_pobj_strategy import flip_prep_pobj_strategy
from .back_and_forth_strategy import back_and_forth_strategy

__all__ = [
    'leaf_strategy',
    'leaf_no_prep_strategy',
    'adverbial_strategy',
    'compound_strategy',
    'flip_prep_pobj_strategy',
    'back_and_forth_strategy',
]
