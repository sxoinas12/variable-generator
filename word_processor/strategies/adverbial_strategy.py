from spacy.tokens import Token
from word_processor.strategies.leaf_strategy import dfs

from word_processor.types import DEP_TYPES


def adverbial_strategy(doc) -> [[Token]]:
    """
    Should return an arrays of variable names based on adverbial strategy
    When a verb appears, if it has children in both directions, then remove the verb and
    add the a left child and a continue to the right child. Does so for all combinations (left,right)

    :param doc: spacy document
    :return Array of strings
    """
    suggestions = []
    for token in doc:
        if token.dep_ == DEP_TYPES['ROOT']:
            suggestions = process_adverbial_clauses(token)
            break

    return suggestions


INVALID_DEP = ['aux', 'prep']
INVALID_POS = ['DET', 'AUX', 'ADP']


def is_valid(node):
    return not (node.pos_ in INVALID_POS) and not (node.dep_ in INVALID_DEP)


# todo - this is dfs
def process_adverbial_clauses(node, result=None, output=None):
    if output is None:
        output = []
    if result is None:
        result = []

    has_children = (node.n_lefts + node.n_rights) > 0
    has_both_directions = node.n_lefts > 0 and node.n_rights > 0

    if node.pos_ == 'VERB' and has_both_directions:
        for lefty in node.lefts:
            for righty in node.rights:
                if is_valid(lefty):
                    process_adverbial_clauses(righty, [*result, lefty], output)
                # if is_valid(righty):
                #     process_adverbial_clauses(lefty, [*result, righty], output)
    elif has_children:
        for u in node.children:
            valid_results = [*result, node] if is_valid(node) else result
            process_adverbial_clauses(u, valid_results, output)
    else:
        valid_results = [*result, node] if is_valid(node) else result
        output.append(valid_results)

    return output
