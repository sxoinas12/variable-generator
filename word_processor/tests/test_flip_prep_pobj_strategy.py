import json

import pytest

import spacy

from word_processor.generators import Generator
from word_processor.strategies import flip_prep_pobj_strategy

nlp = spacy.load("en_core_web_trf")


@pytest.mark.parametrize('text', [
    'allow chat communication for authenticated users',
    'Regex for redacted phone numbers with extra info for PlayStation',
    'generate an authentication token for offline users',
    'serializer for contact roles',
    'Stick element after scroll has passed the element',
])
def test_flip_prep_pobj_strategy(snapshot, text):
    generator = Generator(text=text, strategy=flip_prep_pobj_strategy)

    results = generator.suggest()

    text_results = []
    for suggestion in results:
        text_results.append([token.lemma_ for token in suggestion])

    assert text_results == snapshot
