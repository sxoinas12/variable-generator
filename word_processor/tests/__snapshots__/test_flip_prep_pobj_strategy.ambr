# name: test_flip_prep_pobj_strategy[Regex for redacted phone numbers with extra info for PlayStation]
  <class 'list'> [
    <class 'list'> [
      'redact',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'redact',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'redact',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'redact',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'redact',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'phone',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'phone',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'phone',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'phone',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'phone',
      'redact',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'phone',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'extra',
      'info',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'number',
      'regex',
    ],
    <class 'list'> [
      'PlayStation',
      'info',
      'number',
      'regex',
    ],
  ]
---
# name: test_flip_prep_pobj_strategy[Stick element after scroll has passed the element]
  <class 'list'> [
    <class 'list'> [
      'stick',
      'element',
    ],
    <class 'list'> [
      'stick',
      'after',
      'element',
    ],
    <class 'list'> [
      'stick',
      'scroll',
      'element',
    ],
  ]
---
# name: test_flip_prep_pobj_strategy[allow chat communication for authenticated users]
  <class 'list'> [
    <class 'list'> [
      'allow',
      'chat',
      'communication',
    ],
    <class 'list'> [
      'authenticate',
      'user',
      'allow',
    ],
  ]
---
# name: test_flip_prep_pobj_strategy[generate an authentication token for offline users]
  <class 'list'> [
    <class 'list'> [
      'generate',
      'authentication',
      'token',
    ],
    <class 'list'> [
      'offline',
      'user',
      'generate',
      'authentication',
      'token',
    ],
    <class 'list'> [
      'offline',
      'user',
      'generate',
      'token',
    ],
  ]
---
# name: test_flip_prep_pobj_strategy[serializer for contact roles]
  <class 'list'> [
    <class 'list'> [
      'contact',
      'role',
      'serializer',
    ],
  ]
---
